CRM/Core/Payment/BaseIPN.php

```diff
@@ -214,7 +214,7 @@ public function loadObjects($input, &$ids, &$objects, $required, $paymentProcess
    * @return bool
    * @throws \CiviCRM_API3_Exception
    */
-  public function failed(&$objects, &$transaction, $input = []) {
+  public function failed(&$objects, $transaction = NULL, $input = []) {
```

Comment:
We could instantiate a transaction within this function - but I think in the case of a crash not updating the contribution to failed would likely be worse than a partial update